ngtests: Test fetch target

sechkova opened the pull request:
Fixes #1642 

**Description of the changes being introduced by the pull request**:

Adds tests for 'Fetch target' from 'Detailed client workflow' and target files storing/loading from cache.


**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

sechkova committed:
Add tests/test_updater_fetch_target.py

Add a new test file and class for testing target files
fetching.
Move test_targets from test_updater_with_simulator.py to
tests/test_updater_fetch_targets.py.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Rename and simplify test_targets

Remove parts of the test case which are covered in other
tests, this way making its purpose clearer.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add test_invalid_target*

Add test cases covering downloading and loading from cache
targets with non-matching hash and length.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

kairoaraujo commented on tests/test_updater_fetch_target.py:
```diff
@@ -0,0 +1,166 @@
+        updater = self._init_updater()
+        # target now exists, is not in cache yet
+        info = updater.get_targetinfo(targetpath)
+        assert info is not None
```
```suggestion
        self.assertIsNotNone(info)
```

jku replied:
I think these are for mypy: otherwise you'd need something like 

```
if info is None:
    self.fail("unexpected get_targetinfo() failure")
else:
    # test code using info
```

`assertIsNotNone` is not understood by mypy but `assert` is: then you can avoid the indentation for the rest of the test code

sechkova replied:
What Jussi said :)

kairoaraujo commented on tests/test_updater_fetch_target.py:
```diff
@@ -0,0 +1,166 @@
+
+        updater = self._init_updater()
+        info = updater.get_targetinfo(targetpath)
+        assert info is not None
```
```suggestion
        self.assertIsNotNone(info)
```

kairoaraujo commented on tests/test_updater_fetch_target.py:
```diff
@@ -0,0 +1,166 @@
+        # Download the target
+        updater = self._init_updater()
+        info = updater.get_targetinfo(targetpath)
+        assert info is not None
```
```suggestion
        self.assertIsNotNone(info)
```

kairoaraujo commented on tests/test_updater_fetch_target.py:
```diff
@@ -0,0 +1,166 @@
+        # Newer content is detected, old cached version is not used
+        updater = self._init_updater()
+        info = updater.get_targetinfo(targetpath)
+        assert info is not None
```
```suggestion
        self.assertIsNotNone(info)
```

kairoaraujo reviewed (commented):
I have just a small suggestion.

jku commented on tests/test_updater_fetch_target.py:
```diff
@@ -0,0 +1,166 @@
+        self.temp_dir.cleanup()
+
+    def _init_updater(self) -> Updater:
+        """Creates a new updater and runs refresh."""
```
does not run refresh

jku commented on tests/test_updater_fetch_target.py:
```diff
@@ -0,0 +1,166 @@
+        return updater
+
+    targets: utils.DataSet = {
+        "standard case": ("targetpath", b"content", "targetpath"),
```
I think it would be good to add at least a comment about the contents of this tuple (or make it into a namedtuple or dataclass). Understanding what each "targetpath" means currently requires reading code.

sechkova replied:
I added a data class for target files in another PR before this one and I was planning to use it but it is not yet merged so in other words I got tangled in my own PRs ...
https://github.com/theupdateframework/python-tuf/blob/a0e74c1dbf2ca3b76f32516fda1b71ce01e89db6/tests/test_updater_delegation_graphs.py#L40

sechkova replied:
I added a dataclass `TestTarget` in `test_updater_fetch_target.py` too which is slightly different than the one in `test_updater_delegation_graphs.py` . I realise the same name is not great but I was wondering should I rename one of the classes or we should try making the dataclasses reusable between test files. d1bc201

jku commented on tests/test_updater_fetch_target.py:
```diff
@@ -0,0 +1,166 @@
+        self.sim.add_target("targets", b"contentv2", targetpath)
+        self.sim.update_snapshot()
+
+        # Newer content is detected, old cached version is not used
+        updater = self._init_updater()
+        info = updater.get_targetinfo(targetpath)
+        assert info is not None
+        self.assertIsNone(updater.find_cached_target(info))
```
please continue with download_target and verifying that
* find_cached_target() now returns the path
* cached files now has the new content

jku reviewed (changes requested):
Looks good. I think the cache-miss case should test that we successfully overwrite the "incorrect" file as well. Other comments are just nits.

It feels like three test cases is not a lot... but I can't think of what else to add so :shrug:

Would using delegated targets give us any useful test coverage? the actual delegation lookups are tested elsewhere so maybe not?

sechkova committed:
Define a TestTarget dataclass

Use a dataclass for a better representation of
the target files in the test data set.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
The cache miss test is improved and I agree defining a dataclass does improve readability I am not sure about the naming though, see my comment above.

jku reviewed (approved):
Looks good to me!

kairoaraujo commented:
LGTM
